### PR TITLE
Fix a bug where two threads access MPSC queue

### PR DESCRIPTION
**Pull request by minwoox**
Motivation:
We use MPSC queue in `DefaultStreamMessage`. If more than one thread consumes the MPSC queue,
the index of the queue is not correctly updated so a thread might spin forever in `peek()` as we've seen in #2683.
This bug was introduced from 0.98.4 #2539 by me. 😢 

Modifications:
- Fix to consume the MPSC queue only by the subscription thread.

Result:
- You no longer see the spinning thread in `StreamMessage`.

### CODE REVIEW

File: core/src/main/java/com/linecorp/armeria/common/stream/DefaultStreamMessage.java [hunk 1]
```diff
@@ -174,25 +174,33 @@ private void abort0(Throwable cause) {
...
             }
+        }
+    }
 
-            notifySubscriberOfCloseEvent(subscription, newCloseEvent(cause));
+    private void abort0(Throwable cause, SubscriptionImpl subscription) {
+        final Object o = queue.peek();
+        // If there's no data pushed (i.e empty stream), notify subscriber with the event pushed by
+        // close() or close(cause).
+        if (!wroteAny && o instanceof AbstractStreamMessage.CloseEvent) {
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/common/stream/DefaultStreamMessage.java:198]: nit:
```suggestion
        if (!wroteAny && o instanceof CloseEvent) {
```

Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/common/stream/DefaultStreamMessage.java]: Fixed. Thanks!

**Comment by trustin** [comment]
Any chance we can write a test case for this?

**Comment by trustin** [comment]
Thanks, @minwoox

**Pull request merged**